seatunnel-core/seatunnel-core-base/src/main/java/org/apache/seatunnel/config/PluginFactory.java

```diff
@@ -70,16 +82,62 @@
         PLUGIN_BASE_CLASS_MAP.put(EngineType.FLINK, flinkBaseClassMap);
     }
 
-    public PluginFactory(Config config, EngineType engineType) {
+    public PluginFactory(Config config, EngineType engineType, String seaTunnelHome) {
         this.config = config;
         this.engineType = engineType;
+        this.seaTunnelHome = seaTunnelHome;
+        this.pluginJarPaths = searchPluginJar();
+        this.defaultClassLoader = initClassLoaderWithPaths(this.pluginJarPaths);
+    }
+
+    private ClassLoader initClassLoaderWithPaths(List<URL> pluginJarPaths) {
+        return new URLClassLoader(pluginJarPaths.toArray(new URL[0]),
+                Thread.currentThread().getContextClassLoader());
+    }
+
+    @Nonnull
+    private List<URL> searchPluginJar() {
+
+        File pluginDir = new File(this.seaTunnelHome + "/" + PLUGIN_DIR_NAME + "/" + this.engineType.getEngine());
+        if (!pluginDir.exists() || pluginDir.listFiles() == null) {
+            return new ArrayList<>();
+        }
+        File[] plugins =
+                Arrays.stream(pluginDir.listFiles()).filter(f -> f.getName().endsWith(".jar")).toArray(File[]::new);
+        return Arrays.stream(PluginType.values()).flatMap(type -> {
+            List<URL> pluginList = new ArrayList<>();
+            List<? extends Config> configList = config.getConfigList(type.getType());
+            configList.forEach(pluginConfig -> {
+                try {
+                    for (File plugin : plugins) {
+                        try {
+                            createPluginInstanceIgnoreCase(type, pluginConfig.getString(PLUGIN_NAME_KEY),
+                                    new URLClassLoader(new URL[]{plugin.toURI().toURL()},
+                                            Thread.currentThread().getContextClassLoader()));
+                            pluginList.add(plugin.toURI().toURL());
+                            break;
+                        } catch (ClassNotFoundException ignored) {
+                            // do nothing
+                        }
+
+                    }
+                } catch (Exception e) {
+                    throw new RuntimeException(e);
+                }
+            });
+            return pluginList.stream();
+        }).collect(Collectors.toList());
+    }
```

Comment:
I don't think we need to load plugin class and create instance in this part.
This part is just to find which jar need to be used in this job. We don't need to load all jars, if we load all class, there may still exist some class conflict issue. We already know the mapping of plugin to plugin jar, so we can easily to find the jar path.

Reply:
> I don't think we need to load plugin class and create instance in this part. This part is just to find which jar need to be used in this job. We don't need to load all jars, if we load all class, there may still exist some class conflict issue. We already know the mapping of plugin to plugin jar, so we can easily to find the jar path.



> @BenJFan I still doubt that this way can work. Since we still use `CliFrontend` to submit our Flink job, so the main class defined in `SeatunnelFlink` will be executed in Flink master side, not in our client side. Don't we need to overwrite the `CliFrontend`?

At now, we use flink client mode, don't need overwrite it, but if we use cluster mode in the future, maybe we should consider about it.